src/Tests/InMemoryRepositorySpec.cs

```diff
@@ -0,0 +1,132 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Mqtt.Sdk.Storage;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Tests
+{
+	public class InMemoryRepositorySpec
+    {
+		[Fact]
+		public void when_creating_item_then_succeeds()
+		{
+			var repository = new InMemoryRepository<FooStorageObject>();
+
+			repository.Create(new FooStorageObject { Id = "Foo1", Value = 1 });
+			repository.Create(new FooStorageObject { Id = "Foo2", Value = 2 });
+			repository.Create(new FooStorageObject { Id = "Foo3", Value = 3 });
+			repository.Create(new FooStorageObject { Id = "Foo4", Value = 4 });
+
+			Assert.Equal(4, repository.GetAll().Count());
+		}
+
+		[Fact]
+		public void when_updating_item_then_succeeds()
+		{
+			var repository = new InMemoryRepository<FooStorageObject>();
+			var item = new FooStorageObject { Id = "Foo1", Value = 1 };
+
+			repository.Create(item);
+
+			item.Value = 2;
+
+			repository.Update(item);
+			
+			Assert.Equal(2, repository.GetAll().First().Value);
+		}
+
+		[Fact]
+		public void when_deleting_item_then_succeeds()
+		{
+			var repository = new InMemoryRepository<FooStorageObject>();
+			var item = new FooStorageObject { Id = "Foo1", Value = 1 };
+
+			repository.Create(item);
+			repository.Delete(item.Id);
+
+			Assert.Empty(repository.GetAll());
+		}
+
+		[Fact]
+		public void when_deleting_item_by_id_then_succeeds()
+		{
+			var repository = new InMemoryRepository<FooStorageObject>();
+			var item = new FooStorageObject { Id = "Foo1", Value = 1 };
+
+			repository.Create(item);
+			repository.Delete("Foo1");
+
+			Assert.Empty(repository.GetAll());
+		}
+
+		[Fact]
+		public void when_deleting_item_with_invalid_id_then_does_not_delete()
+		{
+			var repository = new InMemoryRepository<FooStorageObject>();
+			var item = new FooStorageObject { Id = "Foo1", Value = 1 };
+
+			repository.Create(item);
+			repository.Delete("Foo2");
+
+			Assert.NotEmpty(repository.GetAll());
+		}
+
+		[Fact]
+		public void when_checking_existing_items_with_expressions_then_succeeds()
+		{
+			var repository = new InMemoryRepository<FooStorageObject>();
+
+			repository.Create(new FooStorageObject { Id = "Foo1", Value = 1 });
+			repository.Create(new FooStorageObject { Id = "Foo2", Value = 2 });
+			repository.Create(new FooStorageObject { Id = "Foo3", Value = 3 });
+			repository.Create(new FooStorageObject { Id = "Foo4", Value = 4 });
+
+			Assert.Equal(4, repository.GetAll().Where(x => x.Id.StartsWith("Foo")).Count());
```

Comment:
There aren't any more expressions, just a linq query over the resulting enumerable, right? If so, this test should die (and related ones)